File: frontend/src/types/incident.ts
// Shared TypeScript interfaces for the Incident domain (Part 1)

export type IncidentType =
  | 'fire'
  | 'flood'
  | 'medical'
  | 'road_accident'
  | 'building_collapse'
  | 'electrical_hazard'
  | 'missing_person'
  | 'trapped_person'
  | 'security_emergency'
  | 'road_blockage'
  | 'other';

export type IncidentPriority = 'low' | 'medium' | 'high' | 'critical';

export type IncidentStatus =
  | 'reported'
  | 'acknowledged'
  | 'verified'
  | 'assigned'
  | 'accepted'
  | 'en_route'
  | 'on_scene'
  | 'resolving'
  | 'resolved'
  | 'rejected'
  | 'false_alarm'
  | 'duplicate'
  | 'cancelled'
  | 'needs_information';

export interface Incident {
  id: number;
  incident_number: string;
  title: string;
  description: string | null;
  type: IncidentType;
  status: IncidentStatus;
  priority: IncidentPriority;
  calculated_priority: IncidentPriority | null;
  user_selected_priority: IncidentPriority | null;
  severity_score: number | null;
  priority_reason: string | null;
  latitude: number | null;
  longitude: number | null;
  address: string | null;
  details: Record<string, unknown> | null;
  resource_recommendations: string[];
  reporter_id: number;
  created_at: string;
  updated_at: string;
  // Phase 3
  possible_duplicate: boolean;
  duplicate_of_incident_id: number | null;
  reported_at: string | null;
  acknowledged_at: string | null;
  verified_at: string | null;
  assigned_at: string | null;
  accepted_at: string | null;
  en_route_at: string | null;
  on_scene_at: string | null;
  resolving_at: string | null;
  resolved_at: string | null;
}

export interface IncidentTimelineEntry {
  status: IncidentStatus;
  changed_at: string;
  reason: string | null;
}

export interface IncidentListResponse {
  incidents: Incident[];
  total: number;
}

export interface CreateIncidentPayload {
  title: string;
  type: IncidentType;
  description?: string;
  latitude?: number;
  longitude?: number;
  address?: string;
  details?: Record<string, unknown>;
  user_selected_priority?: IncidentPriority;
}

export interface UpdateIncidentStatusPayload {
  status: IncidentStatus;
  reason?: string;
}

// WS incident event envelope
export interface IncidentSocketEvent {
  type: 'incident';
  event: 'created' | 'status_updated';
  payload: Incident;
}

// WS assignment/responder events
export interface AssignmentSocketEvent {
  type: 'incident';
  event: 'incident_assigned' | 'assignment_updated' | 'responder_status_updated';
  payload: Record<string, unknown>;
}

export type AssignmentStatus = 'pending' | 'accepted' | 'rejected' | 'cancelled' | 'completed';

export interface Assignment {
  id: number;
  incident_id: number;
  responder_id: number;
  assigned_by: number;
  assigned_at: string;
  status: AssignmentStatus;
  accepted_at: string | null;
  rejected_at: string | null;
  completed_at: string | null;
  rejection_reason: string | null;
  notes: string | null;
  incident?: Incident;
}

export interface Responder {
  id: number;
  email: string;
  role: string;
  availability: 'AVAILABLE' | 'BUSY' | 'OFFLINE';
  latitude: number | null;
  longitude: number | null;
  active_assignments: number;
  distance_km?: number | null;
}

File: backend/app/schemas/incident.py
from datetime import datetime
from typing import List, Optional, Any
from pydantic import BaseModel, Field, validator
from enum import Enum

class IncidentType(str, Enum):
    fire = "fire"
    flood = "flood"
    medical = "medical"
    other = "other"
    road_accident = "road_accident"
    building_collapse = "building_collapse"
    electrical_hazard = "electrical_hazard"
    missing_person = "missing_person"
    trapped_person = "trapped_person"
    security_emergency = "security_emergency"
    road_blockage = "road_blockage"
    # add more as needed

class IncidentPriority(str, Enum):
    low = "low"
    medium = "medium"
    high = "high"
    critical = "critical"

class IncidentStatus(str, Enum):
    reported = "reported"
    acknowledged = "acknowledged"
    verified = "verified"
    assigned = "assigned"
    accepted = "accepted"
    en_route = "en_route"
    on_scene = "on_scene"
    resolving = "resolving"
    resolved = "resolved"
    rejected = "rejected"
    false_alarm = "false_alarm"
    duplicate = "duplicate"
    cancelled = "cancelled"
    needs_information = "needs_information"

class CreateIncident(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=2000)
    type: IncidentType
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    address: Optional[str] = None
    # Dynamic questionnaire answers – generic JSON container
    details: Optional[Any] = None
    # Optional user‑selected priority (can be omitted)
    user_selected_priority: Optional[IncidentPriority] = None

    @validator("latitude")
    def lat_range(cls, v):
        if v is not None and not (-90 <= v <= 90):
            raise ValueError("latitude must be between -90 and 90")
        return v

    @validator("longitude")
    def lon_range(cls, v):
        if v is not None and not (-180 <= v <= 180):
            raise ValueError("longitude must be between -180 and 180")
        return v

class UpdateIncident(BaseModel):
    title: Optional[str] = Field(None, max_length=200)
    description: Optional[str] = Field(None, max_length=2000)
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    address: Optional[str] = None
    details: Optional[Any] = None
    user_selected_priority: Optional[IncidentPriority] = None

    @validator("latitude")
    def lat_range(cls, v):
        if v is not None and not (-90 <= v <= 90):
            raise ValueError("latitude must be between -90 and 90")
        return v

    @validator("longitude")
    def lon_range(cls, v):
        if v is not None and not (-180 <= v <= 180):
            raise ValueError("longitude must be between -180 and 180")
        return v

class IncidentStatusUpdate(BaseModel):
    status: IncidentStatus
    reason: Optional[str] = None

class IncidentStatusHistoryResponse(BaseModel):
    status: IncidentStatus
    changed_at: datetime
    reason: Optional[str] = None

    class Config:
        from_attributes = True

class IncidentResponse(BaseModel):
    incident_number: str
    title: str
    description: Optional[str]
    type: IncidentType
    status: IncidentStatus
    priority: IncidentPriority
    latitude: Optional[float]
    longitude: Optional[float]
    address: Optional[str]
    reporter_id: int
    created_at: datetime
    updated_at: datetime
    details: Optional[Any]
    resource_recommendations: List[str] = []
    severity_score: Optional[int] = None
    calculated_priority: IncidentPriority
    user_selected_priority: Optional[IncidentPriority] = None

    class Config:
        from_attributes = True

class IncidentListResponse(BaseModel):
    incidents: List[IncidentResponse]

    class Config:
        from_attributes = True
